File: backend/validation.py
from typing import List

from .models import FlowConfig


class FlowValidator:
    def __init__(self, flow: FlowConfig):
        self.flow = flow
        self.errors: List[str] = []

    def validate(self) -> List[str]:
        """Run all validation checks and return list of errors"""
        self.errors = []

        self._validate_initial_node()
        self._validate_node_references()
        self._validate_node_contents()

        return self.errors

    def _validate_initial_node(self):
        """Ensure initial node exists and is valid"""
        if not self.flow.initial_node:
            self.errors.append("Initial node must be specified")
        elif self.flow.initial_node not in self.flow.nodes:
            self.errors.append(f"Initial node '{self.flow.initial_node}' not found in nodes")

    def _validate_node_references(self):
        """Ensure all function references point to valid nodes"""
        for node_id, node in self.flow.nodes.items():
            function_names = {
                func["function"]["name"]
                for func in node.functions
                if "function" in func and "name" in func["function"]
            }

            # Check if function names reference valid nodes
            for func_name in function_names:
                if func_name not in self.flow.nodes and func_name != "end":
                    self.errors.append(
                        f"Node '{node_id}' has function '{func_name}' that doesn't reference a valid node"
                    )

    def _validate_node_contents(self):
        """Validate individual node configurations"""
        for node_id, node in self.flow.nodes.items():
            # Validate messages
            if not node.messages:
                self.errors.append(f"Node '{node_id}' must have at least one message")

            for msg in node.messages:
                if "role" not in msg:
                    self.errors.append(f"Message in node '{node_id}' missing 'role'")
                if "content" not in msg:
                    self.errors.append(f"Message in node '{node_id}' missing 'content'")

            # Validate functions
            for func in node.functions:
                if "function" not in func:
                    self.errors.append(f"Function in node '{node_id}' missing 'function' object")
                elif "name" not in func["function"]:
                    self.errors.append(f"Function in node '{node_id}' missing 'name'")


def validate_flow(flow: FlowConfig) -> List[str]:
    """Convenience function to validate a flow configuration"""
    validator = FlowValidator(flow)
    return validator.validate()

File: js/validation.js
class FlowValidator {
  constructor(flowConfig) {
    this.flow = flowConfig;
    this.errors = [];
  }

  validate() {
    this.errors = [];

    this._validateInitialNode();
    this._validateNodeReferences();
    this._validateNodeContents();

    return this.errors;
  }

  _validateInitialNode() {
    if (!this.flow.initial_node) {
      this.errors.push('Initial node must be specified');
    } else if (!this.flow.nodes[this.flow.initial_node]) {
      this.errors.push(
        `Initial node '${this.flow.initial_node}' not found in nodes`
      );
    }
  }

  _validateNodeReferences() {
    Object.entries(this.flow.nodes).forEach(([nodeId, node]) => {
      if (node.functions) {
        const functionNames = node.functions
          .map((func) => func.function?.name)
          .filter(Boolean);

        functionNames.forEach((funcName) => {
          if (!this.flow.nodes[funcName] && funcName !== 'end') {
            this.errors.push(
              `Node '${nodeId}' has function '${funcName}' that doesn't reference a valid node`
            );
          }
        });
      }
    });
  }

  _validateNodeContents() {
    Object.entries(this.flow.nodes).forEach(([nodeId, node]) => {
      // Validate messages
      if (!node.messages || node.messages.length === 0) {
        this.errors.push(`Node '${nodeId}' must have at least one message`);
      }

      node.messages?.forEach((msg) => {
        if (!msg.role) {
          this.errors.push(`Message in node '${nodeId}' missing 'role'`);
        }
        if (!msg.content) {
          this.errors.push(`Message in node '${nodeId}' missing 'content'`);
        }
      });

      // Validate functions
      node.functions?.forEach((func) => {
        if (!func.function) {
          this.errors.push(
            `Function in node '${nodeId}' missing 'function' object`
          );
        } else if (!func.function.name) {
          this.errors.push(`Function in node '${nodeId}' missing 'name'`);
        }
      });
    });
  }
}

// Helper function to validate a flow configuration
function validateFlow(flowConfig) {
  const validator = new FlowValidator(flowConfig);
  return {
    valid: validator.validate().length === 0,
    errors: validator.errors,
  };
}

// Example usage in main.js:
document.getElementById('export-flow').onclick = async () => {
  try {
    const flowConfig = generateFlowConfig(graph);

    // Validate before export
    const validation = validateFlow(flowConfig);
    if (!validation.valid) {
      console.error('Flow validation errors:', validation.errors);
      if (!confirm('Flow has validation errors. Export anyway?')) {
        return;
      }
    }

    // Continue with export...
    console.log('Generated Flow Configuration:');
    console.log(JSON.stringify(flowConfig, null, 2));
    // ... rest of export code ...
  } catch (error) {
    console.error('Error generating flow configuration:', error);
    alert('Error generating flow configuration: ' + error.message);
  }
};

// Also use during import
document.getElementById('import-flow').onclick = () => {
  const input = document.createElement('input');
  input.type = 'file';
  input.accept = '.json';
  input.onchange = (e) => {
    const file = e.target.files[0];
    const reader = new FileReader();
    reader.onload = (event) => {
      try {
        const flowConfig = JSON.parse(event.target.result);

        // Validate imported flow
        const validation = validateFlow(flowConfig);
        if (!validation.valid) {
          console.error('Flow validation errors:', validation.errors);
          if (!confirm('Imported flow has validation errors. Import anyway?')) {
            return;
          }
        }

        createFlowFromConfig(graph, flowConfig);
        console.log('Successfully imported flow configuration');
      } catch (error) {
        console.error('Error importing flow:', error);
        alert('Error importing flow: ' + error.message);
      }
    };
    reader.readAsText(file);
  };
  input.click();
};
